# Problem Statement 4

# Billboard Distance Optimization Problem

You decide to invest in a series of billboards along Interstate 10 to advertise your stylish new chicken restaurant, Chic-fil-A 🐔.

* **First Set:** Three billboards evenly spaced starting from mile marker 17 and ending on mile marker 28.
* **Second Set:** Three billboards evenly spaced starting on mile marker 32 and ending on mile marker 36.
* **Ad Sequence (from 17 to 36):** A, B, C, C, B, A.
* **Restaurant Location:** Mile marker 30.

**Goal:** Determine how far each "B" ad is from your restaurant.



# Solution 

In [1]:
import numpy as np

In [2]:
# Step 1: First group of billboards (mile 17 to 28), 3 evenly spaced
group1=np.linspace(17,28,3)

# Step 2: Second group of billboards (mile 32 to 36), 3 evenly spaced
group2=np.linspace(32,36,3)

In [3]:
# Step 3: Combine into one array of all 6 mile markers, in order
mile_markers=np.concatenate([group1,group2])
print('mile_markers',mile_markers)

mile_markers [17.  22.5 28.  32.  34.  36. ]


In [4]:
# Step 4: The ads in order
ads=np.array(['A','B','C','C','B','A'])
print( ' ads    ',ads)

 ads     ['A' 'B' 'C' 'C' 'B' 'A']


In [5]:
# Step 5: Find where the ad is 'B'
b_mask=(ads=='B')
b_mile_markers=mile_markers[b_mask]
print('B mile markers:',b_mile_markers)

B mile markers: [22.5 34. ]


In [6]:
# Step 6: Restaurant location
restaurant=30

In [7]:
# Step 7: Distance of each B billboard from the restaurant
distances = np.abs(b_mile_markers - restaurant)
print("Distances of B billboards from restaurant:", distances)


Distances of B billboards from restaurant: [7.5 4. ]


### Breaking down the key lines
* np.linspace(17, 28, 3)\
Generates 3 evenly spaced numbers between 17 and 28 inclusive — exactly what "3 billboards starting at 17, ending at 28" means. Result: [17, 22.5, 28].
* np.concatenate([group1, group2])\
Joins the two arrays end-to-end into one array of 6 mile markers, preserving order.
ads == 'B'\
This compares every element of the ads array to the string 'B', producing a boolean mask — an array of True/False values, one per position: [False, True, False, False, True, False].
* mile_markers[b_mask]\
Using that boolean mask to index mile_markers pulls out only the mile markers where the mask is True — i.e., only the positions where the ad is 'B'. This is called boolean indexing, and it's one of NumPy's most powerful features.
* np.abs(b_mile_markers - restaurant)\
Subtracts 30 from each B mile marker, then takes the absolute value of each result, so direction (before/after the restaurant) doesn't matter — only distance does.

# np.abs

* np.abs() — Absolute Value, Element-wise
* np.abs() is NumPy's function for computing the absolute value of numbers — meaning it strips away any negative sign, turning every value into its non-negative version.